src/sinks/datadog/traces/stats.rs

```diff
@@ -398,36 +406,107 @@ impl Aggregator {
         });
         m
     }
+
+    /// Flushes the bucket cache of stale entries.
+    /// Based on https://github.com/DataDog/datadog-agent/blob/cfa750c7412faa98e87a015f8ee670e5828bbe7f/pkg/trace/stats/concentrator.go#L38-L41
+    /// It means that we cache and can compute stats only for the last `BUCKET_WINDOW_LEN * BUCKET_DURATION_NANOSECONDS` and after such time,
+    /// buckets are then flushed. This only applies to past buckets. Stats buckets in the future are cached with no restriction.
+    fn flush(&mut self) {
+        let now = Utc::now().timestamp_nanos() as u64;
+        let flush_cutoff_time = now - (BUCKET_DURATION_NANOSECONDS * BUCKET_WINDOW_LEN);
+
+        // remove entries from the map if the start time is outside the retaining window
+        self.buckets.retain(|&bucket_start, _bucket| {
+            let flush = bucket_start >= flush_cutoff_time;
```

Comment:
This logic feels a little funny to me.

Assume we create bucket #1 at t=45, where bucket duration is 30. That bucket gets a `start` time of 30 (`45 - (45 % 30)`), which means it should last until 60 (`30 + 30`). We create bucket #2 after that, following the same logic, with a `start` of 60, lasting until 90.

We call `flush` at t=95, where `flush_cutoff_time` will be 35 (`95 - (30 * 2)`). When evaluating against bucket #1, `flush` would be `false`, as 30 is not greater than or equal to 35... but we're dropping half of the data because the bucket's start was just over the line.

Essentially, do we want to ensure we never have _more_ than X seconds of total data? If so, this seems fine... but it does feel weird that we might be oscillating between providing anywhere from the last 10 seconds to the last 20 seconds of data (using the real constants) depending on when we flush.

I'm not sure if that's also how the agent works, in which case... *shrug*

Reply:
> Essentially, do we want to ensure we never have more than X seconds of total data?

That is my _limited_ understanding from reading the Agent's code.

One thing I detected when I took another look at that logic just now, is that our computation of the `bucket_start` was actually different than the agent's code. The agent code computed the beginning of the window based on the "end" of the span duration (start + end). Here we were only taking the start time into consideration.
So, I changed that.

I also added the missing logic of "oldest timestamp" that we were missing...

Between those changes and correcting the flush/retain logic, let me know if that makes this look a little better.

I'll admit to not fully understanding algorithm and the assumptions on the consumer side of the stats payloads.
The confidence in these changes is largely based on the A/B test with the DD UI *shrug*

Reply:
Attempted to address in https://github.com/vectordotdev/vector/pull/14694/commits/82d0e97b9e3e2b9aae645696f77571971564e4f4